backend/canisters/bucket/impl/src/model/stable_blob_storage.rs

```diff
@@ -0,0 +1,147 @@
+use serde::{Deserialize, Serialize};
+use std::borrow::Cow;
+use std::mem::size_of;
+use types::Hash;
+
+const MAX_VALUE_SIZE: usize = 4 * 1024; // 4KB
+
+type Memory = VirtualMemory<DefaultMemoryImpl>;
+
+#[derive(Serialize, Deserialize)]
+pub struct StableBlobStorage {
+    #[serde(skip, default = "init_blobs")]
+    blobs: StableBTreeMap<Memory, Key, Vec<u8>>,
+}
+
+impl StableBlobStorage {
+    pub fn get(&self, hash: &Hash) -> Option<Vec<u8>> {
+        self.value_iterator(hash).map(|i| i.flat_map(|v| v).collect())
+    }
+
+    pub fn data_size(&self, hash: &Hash) -> Option<u64> {
+        self.value_iterator(hash).map(|i| i.map(|v| v.len() as u64).sum())
+    }
+
+    pub fn exists(&self, hash: &Hash) -> bool {
+        self.value_iterator(hash).is_some()
+    }
+
+    pub fn len(&self) -> u64 {
+        self.blobs.len()
+    }
+
+    pub fn insert(&mut self, hash: Hash, value: Vec<u8>) {
+        for (index, chunk) in value.chunks(MAX_VALUE_SIZE).enumerate() {
+            let key = Key::new(hash, index as u32);
+
+            self.blobs.insert(key, chunk.to_vec()).unwrap();
+        }
+    }
+
+    pub fn remove(&mut self, hash: &Hash) -> bool {
+        let keys: Vec<Key> = self.blobs.range(hash.to_vec(), None).map(|(k, _)| k).collect();
+
+        if keys.is_empty() {
+            false
+        } else {
+            for key in keys {
+                self.blobs.remove(&key);
+            }
+            true
+        }
+    }
+
+    // Returns None if no value exists with the given hash, else provides an iterator over the
+    // value's chunks.
+    fn value_iterator(&self, hash: &Hash) -> Option<impl Iterator<Item = Vec<u8>> + '_> {
+        let mut iter = self.blobs.range(hash.to_vec(), None).map(|(_, v)| v);
+
+        let first = iter.next()?;
+
+        Some([first].into_iter().chain(iter))
+    }
+}
+
+fn init_blobs() -> StableBTreeMap<Memory, Key, Vec<u8>> {
+    let memory_manager = MemoryManager::init(DefaultMemoryImpl::default());
+    let memory = memory_manager.get(MemoryId::new(0));
+
+    StableBTreeMap::init(memory, size_of::<Key>() as u32, MAX_VALUE_SIZE as u32)
+}
+
+impl Default for StableBlobStorage {
+    fn default() -> Self {
+        StableBlobStorage { blobs: init_blobs() }
+    }
+}
+
+#[allow(dead_code)]
+#[repr(packed)]
+struct Key {
+    prefix: Hash,
+    chunk_index_bytes: [u8; 4],
+}
+
+impl Key {
+    fn new(prefix: Hash, chunk_index: u32) -> Key {
+        Key {
+            prefix,
+            chunk_index_bytes: chunk_index.to_be_bytes(),
+        }
+    }
+}
+
+impl Storable for Key {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        let bytes = unsafe { std::slice::from_raw_parts((self as *const Key) as *const u8, size_of::<Key>()) };
+
+        Cow::from(bytes)
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        assert_eq!(bytes.len(), size_of::<Key>());
+
+        unsafe { std::ptr::read(bytes.as_ptr() as *const _) }
+    }
```

Comment:
While this looks correct, I'd recommend avoiding unsafe command unless it's really necessary. Maybe you can replace with something like [this](https://github.com/dfinity/bitcoin-canister/blob/master/canister/src/types.rs#L339).

Reply:
I have a test which covers round tripping the Key to bytes and back again.

It may have a negligible performance boost but the fact that it is zero copy makes me happy :)